x-pack/plugins/lens/public/app_plugin/app.tsx

```diff
@@ -309,88 +318,144 @@ export function App({
-    docStorage
-      .save(doc)
-      .then(({ id }) => {
-        // Prevents unnecessary network request and disables save button
-        const newDoc = { ...doc, id };
-        const currentOriginatingApp = state.originatingApp;
-        setState((s) => ({
-          ...s,
-          isSaveModalVisible: false,
-          originatingApp:
-            newlyCreated && !saveProps.returnToOrigin ? undefined : currentOriginatingApp,
-          persistedDoc: newDoc,
-          lastKnownDoc: newDoc,
-        }));
-        if (docId !== id || saveProps.returnToOrigin) {
-          redirectTo(id, saveProps.returnToOrigin, newlyCreated);
+    const addedToLibrary = state.byValueMode && saveToLibrary;
+
+    const newlyCreated =
+      saveProps.newCopyOnSave || addedToLibrary || (!savedObjectId && !state.byValueMode);
+
+    if (state.byValueMode && !saveToLibrary && redirectToOrigin) {
+      await setState((s: LensAppState) => ({ ...s, persistedDoc: doc }));
+      const { savedObjectId: id, type, ...attributes } = doc;
+      redirectToOrigin({ attributes } as LensByValueInput);
+    } else {
+      await checkForDuplicateTitle(
+        {
+          ...doc,
+          copyOnSave: saveProps.newCopyOnSave,
+          lastSavedTitle: lastKnownDoc.title,
+          getEsType: () => 'lens',
+          getDisplayName: () =>
+            i18n.translate('xpack.lens.app.saveModalType', {
+              defaultMessage: 'Lens visualization',
+            }),
+        },
+        saveProps.isTitleDuplicateConfirmed,
+        saveProps.onTitleDuplicate,
+        {
+          savedObjectsClient: core.savedObjects.client,
+          overlays: core.overlays,
         }
-      })
-      .catch((e) => {
-        // eslint-disable-next-line no-console
-        console.dir(e);
-        trackUiEvent('save_failed');
-        core.notifications.toasts.addDanger(
-          i18n.translate('xpack.lens.app.docSavingError', {
-            defaultMessage: 'Error saving document',
-          })
-        );
-        setState((s) => ({ ...s, isSaveModalVisible: false }));
-      });
+      );
+      docStorage
+        .save(doc)
+        .then(({ savedObjectId: newSavedObjectId }) => {
+          // Prevents unnecessary network request and disables save button
+          const newDoc: Document = { ...doc, savedObjectId: newSavedObjectId };
+          const currentOriginatingApp = state.originatingApp;
+          setState((s) => ({
+            ...s,
+            byValueMode: false,
+            isSaveModalVisible: false,
+            originatingApp:
+              newlyCreated && !saveProps.returnToOrigin ? undefined : currentOriginatingApp,
+            persistedDoc: newDoc,
+            lastKnownDoc: newDoc,
+          }));
+          if (saveProps.returnToOrigin && redirectToOrigin) {
+            redirectToOrigin(
+              newlyCreated
+                ? ({ savedObjectId: newSavedObjectId } as LensByReferenceInput)
+                : undefined
```

Comment:
It looks like this is the same arguments as the `cancel` button, but I don't really understand what the use case here is. I think it could use some explanation of when this branch would be triggered.

Reply:
This branch is triggered when you've edited a lens visualization by reference, then saved it and returned. The `savedObjectId` stays the same, so there's no reason to pass the new information back to dashboard.